### PR TITLE
[RISCV] Teach .option arch to support experimental extensions.

### PR DESCRIPTION
**Pull request by yetingk**
Previously `.option arch` denied extenions are not belongs to RISC-V features. But experimental features have experimental- prefix, so `.option arch` can not serve for experimental extension.
This patch uses the features of extensions to identify extension existance.

### CODE REVIEW

**Comment by jrtc27** [comment]
I’m concerned about testing this; experimental extensions are, by policy, transient in LLVM, eventually either being promoted to non-experimental or being dropped, which means churn in this test every time the chosen extension is no longer present and experimental. What are your thoughts there?

**Comment by jrtc27** [comment]
> > Also, does this obey -menable-experimental-extensions?
> 
> IMO, I prefer this feature depends on `-menable-experimental-extensions`.

Yes, I believe it should. My question is whether it currently does, especially given I don’t believe your current patch is testing it?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 1]
```diff
@@ -51,6 +51,9 @@ STATISTIC(RISCVNumInstrsCompressed,
 
 static cl::opt<bool> AddBuildAttributes("riscv-add-build-attributes",
                                         cl::init(false));
+static cl::opt<bool>
+    DisableExperimentalExtension("riscv-disable-experimental-ext",
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:55]: Why disable instead of enable?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 2]
```diff
@@ -2824,8 +2827,12 @@ bool RISCVAsmParser::parseDirectiveOption() {
         break;
       }
 
-      auto Ext = llvm::lower_bound(RISCVFeatureKV, Arch);
-      if (Ext == std::end(RISCVFeatureKV) || StringRef(Ext->Key) != Arch ||
+      std::string &&Feature = RISCVISAInfo::getTargetFeatureForExtension(Arch);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:2830]: Why an rvalue reference?

Comment (topperc) [reply llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:2830]: I think you should just use `std::string Feature`. I think `std::string &&Feature` will just make a reference to a hidden temporary object.

**Comment by topperc** [comment]
Does this disable use of experimental extensions for a `.option arch` in an inline assembly block?

File: clang/lib/Driver/ToolChains/Arch/RISCV.cpp [hunk 3]
```diff
@@ -169,6 +169,9 @@ void riscv::getRISCVTargetFeatures(const Driver &D, const llvm::Triple &Triple,
     Features.push_back("-relax");
   }
 
+  if (!Args.hasArg(options::OPT_menable_experimental_extensions))
+    Features.push_back("+no-experimental-ext");
```
Comment (topperc) [inline clang/lib/Driver/ToolChains/Arch/RISCV.cpp:173]: We also check `OPT_menable_experimental_extensions` in `getArchFeatures` and push a different feature when it set. Do we need 2 extensions that are the opposite of each other?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 4]
```diff
@@ -51,7 +51,6 @@ STATISTIC(RISCVNumInstrsCompressed,
 
 static cl::opt<bool> AddBuildAttributes("riscv-add-build-attributes",
                                         cl::init(false));
-
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:54]: Unrelated change?

File: llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp [hunk 5]
```diff
@@ -2824,8 +2826,12 @@ bool RISCVAsmParser::parseDirectiveOption() {
...
 
-      auto Ext = llvm::lower_bound(RISCVFeatureKV, Arch);
-      if (Ext == std::end(RISCVFeatureKV) || StringRef(Ext->Key) != Arch ||
+      std::string Feature = RISCVISAInfo::getTargetFeatureForExtension(Arch);
+      if (!enableExperimentalExtension() &&
+          StringRef(Feature).starts_with("experimental-"))
+        return Error(Loc, "Unexpected experimental extensions.");
+      auto Ext = llvm::lower_bound(RISCVFeatureKV, Feature);
+      if (Ext == std::end(RISCVFeatureKV) || StringRef(Ext->Key) != Feature ||
           !RISCVISAInfo::isSupportedExtension(Arch)) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:2835]: Do we still need the `RISCVISAInfo::isSupportedExtension(Arch)` check here? Did `RISCVISAInfo::getTargetFeatureForExtension` already take care of it?

**Review by topperc — APPROVED** [review]
LGTM

**Comment by dyung** [comment]
@yetingk, the two tests you added are failing on a couple of build bots:
- https://lab.llvm.org/buildbot/#/builders/164/builds/51995
- https://lab.llvm.org/buildbot/#/builders/258/builds/17867
- https://lab.llvm.org/buildbot/#/builders/91/builds/25943
- https://lab.llvm.org/buildbot/#/builders/280/builds/3159
- https://lab.llvm.org/buildbot/#/builders/283/builds/1636
- https://lab.llvm.org/buildbot/#/builders/139/builds/64910
- https://lab.llvm.org/buildbot/#/builders/259/builds/3993
- https://lab.llvm.org/buildbot/#/builders/285/builds/1756

Can you take a look and revert if you need time to investigate?

**Comment by zmodem** [comment]
6217abc should fix the test failures.

**Comment by pogo59** [comment]
Driver tests are supposed to verify Driver behavior. That is, use `-###` and make sure the correct option is being passed down to the compiler. They should not be compiling anything.

File: clang/test/Driver/riscv-option-arch.c [hunk 6]
```diff
@@ -0,0 +1,7 @@
+// RUN: %clang --target=riscv64 -menable-experimental-extensions -c -o /dev/null %s
```
Comment (jrtc27) [inline clang/test/Driver/riscv-option-arch.c:1]: Why are these clang tests? This is entirely testing llvm/ code. Never mind the weirdness of putting them in Driver/, nor the fact that they’re missing a REQUIRES: riscv-registered-target given they use the backend, they don’t belong in clang/ at all surely?

File: clang/test/Driver/riscv-option-arch.c [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+// RUN: ! %clang --target=riscv64 -c -o /dev/null %s 2>&1 | FileCheck -check-prefixes=CHECK-ERR %s
```
Comment (jrtc27) [inline clang/test/Driver/riscv-option-arch.c:2]: Use not rather than !
